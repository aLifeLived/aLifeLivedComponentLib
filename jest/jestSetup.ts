/* eslint-disable no-undef */
jest.useFakeTimers({ legacyFakeTimers: true }); // Stops reference error from appearing
