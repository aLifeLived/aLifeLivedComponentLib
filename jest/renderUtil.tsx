import React from 'react';
import { ThemeProvider } from '../src/themes/ThemeProvider';
import { render as rtlRender } from '@testing-library/react-native';
import { library } from '@fortawesome/fontawesome-svg-core';
import { iconList } from '../src/themes/systemIcon/IconList';

library.add(iconList);

function render(ui: any, { ...renderOptions } = {}) {
  function Wrapper({ children }: { children: any }) {
    return <ThemeProvider>{children}</ThemeProvider>;
  }

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything for convenience
export * from '@testing-library/react-native';

// override render method
export { render };
