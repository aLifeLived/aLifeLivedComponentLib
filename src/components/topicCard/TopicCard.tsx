import React from 'react';
import { Box } from '../../themes/box/Box';
import { SystemButton } from '../../themes/systemButton/SystemButton';
import { Text } from '../../themes/text/Text';
import { Icon } from '../icon/Icon';

type TopicCardProps = {
  title: string;
  hasUserCompleted: boolean;
  description: string;
  duration: number;
  onPress: () => void;
};

export const TopicCard: React.FC<TopicCardProps> = ({
  title,
  hasUserCompleted,
  description,
  duration,
  onPress,
}) => {
  return (
    <SystemButton onPress={() => onPress()}>
      <Box
        borderColor="borderColor"
        backgroundColor="white"
        borderWidth={1}
        borderRadius="medium"
        margin="m"
        flexDirection="row"
        alignContent="center"
        alignItems="center"
        justifyContent="space-between"
        boxShadow="md"
      >
        <Box marginLeft="s" marginBottom="s" width="80%">
          <Text
            color="textContrastHigh"
            variant="h3"
            marginTop="s"
            fontWeight="bold"
            numberOfLines={1}
            marginRight="m"
          >
            {title}
          </Text>
          <Text marginTop="s" variant="caption" numberOfLines={1}>
            {description}
          </Text>
          <Box flexDirection="row" marginTop="s" alignItems="center">
            <Icon icon={['far', 'clock']} testID="clock-icon" />
            <Text color="primary" variant="caption" marginLeft="s">
              {duration} mins
            </Text>
          </Box>
        </Box>

        {hasUserCompleted && (
          <Icon
            icon={['fas', 'check-circle']}
            color="primary"
            marginRight="m"
            size={20}
            testID="checked-icon"
          />
        )}
      </Box>
    </SystemButton>
  );
};
