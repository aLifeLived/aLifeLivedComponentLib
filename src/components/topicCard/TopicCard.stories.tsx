import React from 'react';
import { storiesOf } from '@storybook/react-native';
import { action } from '@storybook/addon-actions';

// Component
import { TopicCard } from './TopicCard';
import { Box } from '../../themes/box/Box';

storiesOf('Cards', module)
  .addDecorator(Story => Story())
  .add('Default topic card', () => (
    <Box>
      <Box margin="m">
        <TopicCard
          onPress={action('On press fired')}
          title="Ancestry"
          duration={90}
          hasUserCompleted={false}
        />
      </Box>
      <Box margin="m">
        <TopicCard
          onPress={action('On press fired')}
          title="This is a really really long title, here it will break the topic card, if it gets any longer it may break the topic card!!"
          duration={25}
          hasUserCompleted={true}
        />
      </Box>
      <Box margin="m">
        <TopicCard
          onPress={action('On press fired')}
          title="School Years"
          duration={25}
          hasUserCompleted={false}
        />
      </Box>
      <Box margin="m">
        <TopicCard
          onPress={action('On press fired')}
          title="Lessons Learned"
          duration={25}
          hasUserCompleted={true}
        />
      </Box>
    </Box>
  ));
