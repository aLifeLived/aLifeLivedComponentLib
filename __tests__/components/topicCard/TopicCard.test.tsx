import React from 'react';
import { render, fireEvent } from './../../../jest/renderUtil';

// Component
import { TopicCard } from '../../../src/components/topicCard/TopicCard';

const onPress = jest.fn();
const defaultTitle = 'Test Title';

const setup = (hasUserCompleted = false) =>
  render(
    <TopicCard
      title={defaultTitle}
      hasUserCompleted={hasUserCompleted}
      duration={25}
      onPress={onPress}
    />
  );

describe('components >> topic card ', () => {
  beforeEach(() => onPress.mockClear());
  describe('User has NOT completed', () => {
    test('Should render component correctly', () => {
      const { toJSON } = setup();
      expect(toJSON()).toMatchSnapshot();
    });

    test('Should find title and duration', () => {
      const { getByText } = setup();
      expect(getByText(defaultTitle)).toBeDefined();
      expect(getByText('25 mins')).toBeDefined();
    });

    test('Should fire on press', () => {
      const { getByText } = setup();
      fireEvent.press(getByText(defaultTitle));
      expect(onPress).toBeCalledTimes(1);
    });

    test('Should find clock icon', () => {
      const { getByTestId } = setup();
      expect(getByTestId('clock-icon')).toBeDefined();
    });
  });

  describe('User HAS completed', () => {
    const hasUserCompletedTopic = true;
    test('Should render component correctly', () => {
      const { toJSON } = setup(hasUserCompletedTopic);
      expect(toJSON()).toMatchSnapshot();
    });

    test('Should find title', () => {
      const { getByText } = setup(hasUserCompletedTopic);
      expect(getByText(defaultTitle)).not.toBeNull();
    });

    test('Should fire on press', () => {
      const { getByText } = setup(hasUserCompletedTopic);
      fireEvent.press(getByText(defaultTitle));
      expect(onPress).toBeCalledTimes(1);
    });

    test('Should find icons', () => {
      const { getByTestId } = setup(hasUserCompletedTopic);
      expect(getByTestId('checked-icon')).not.toBeNull();
    });
  });
});
