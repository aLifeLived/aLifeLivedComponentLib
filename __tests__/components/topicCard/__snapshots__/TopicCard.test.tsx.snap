// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components >> topic card  User HAS completed Should render component correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#ffffff",
          "borderColor": "#2828281e",
          "borderRadius": 16,
          "borderWidth": 1,
          "display": "flex",
          "paddingBottom": 24,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 24,
        },
        Array [
          Object {
            "elevation": 2,
            "shadowColor": "#282828",
            "shadowOffset": Object {
              "height": 4,
              "width": 0,
            },
            "shadowOpacity": 0.1,
            "shadowRadius": 10,
          },
        ],
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginRight": 8,
          },
          Array [
            Object {},
          ],
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "display": "flex",
              "marginRight": 16,
              "width": "75%",
            },
            Array [
              Object {},
            ],
          ]
        }
      >
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#282828",
                "display": "flex",
                "fontFamily": "SFUIDisplay-Regular",
                "fontSize": 18,
                "lineHeight": 20,
              },
            ]
          }
        >
          Test Title
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "display": "flex",
            },
            Array [
              Object {},
            ],
          ]
        }
      >
        <RNSVGSvgView
          align="xMidYMid"
          bbHeight={20}
          bbWidth={20}
          fill="#67B6A8"
          focusable={false}
          height={20}
          meetOrSlice={0}
          minX={0}
          minY={0}
          secondaryFill={null}
          secondaryOpacity={0.4}
          size={20}
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
                "borderWidth": 0,
              },
              Object {},
              Object {
                "flex": 0,
                "height": 20,
                "width": 20,
              },
            ]
          }
          testID="checked-icon"
          vbHeight={512}
          vbWidth={512}
          width={20}
        >
          <RNSVGGroup
            fill={4284987048}
            propList={
              Array [
                "fill",
              ]
            }
          >
            <RNSVGPath
              d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
              fill={4284987048}
              fillOpacity={1}
              propList={
                Array [
                  "fill",
                  "fillOpacity",
                ]
              }
            />
          </RNSVGGroup>
        </RNSVGSvgView>
      </View>
    </View>
  </View>
</View>
`;

exports[`components >> topic card  User has NOT completed Should render component correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#ffffff",
          "borderColor": "#2828281e",
          "borderRadius": 16,
          "borderWidth": 1,
          "display": "flex",
          "paddingBottom": 24,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 24,
        },
        Array [
          Object {
            "elevation": 2,
            "shadowColor": "#282828",
            "shadowOffset": Object {
              "height": 4,
              "width": 0,
            },
            "shadowOpacity": 0.1,
            "shadowRadius": 10,
          },
        ],
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginRight": 8,
          },
          Array [
            Object {},
          ],
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "display": "flex",
              "marginRight": 16,
              "width": "75%",
            },
            Array [
              Object {},
            ],
          ]
        }
      >
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#282828",
                "display": "flex",
                "fontFamily": "SFUIDisplay-Regular",
                "fontSize": 18,
                "lineHeight": 20,
              },
            ]
          }
        >
          Test Title
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "display": "flex",
            },
            Array [
              Object {},
            ],
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "display": "flex",
                "flexDirection": "row",
              },
              Array [
                Object {},
              ],
            ]
          }
        >
          <RNSVGSvgView
            align="xMidYMid"
            bbHeight={16}
            bbWidth={16}
            fill="#A9ABAD"
            focusable={false}
            height={16}
            meetOrSlice={0}
            minX={0}
            minY={0}
            secondaryFill={null}
            secondaryOpacity={0.4}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "borderWidth": 0,
                },
                Object {
                  "marginRight": 8,
                },
                Object {
                  "flex": 0,
                  "height": 16,
                  "width": 16,
                },
              ]
            }
            testID="clock-icon"
            vbHeight={512}
            vbWidth={512}
            width={16}
          >
            <RNSVGGroup
              fill={4289309613}
              propList={
                Array [
                  "fill",
                ]
              }
            >
              <RNSVGPath
                d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm0 448c-110.5 0-200-89.5-200-200S145.5 56 256 56s200 89.5 200 200-89.5 200-200 200zm61.8-104.4l-84.9-61.7c-3.1-2.3-4.9-5.9-4.9-9.7V116c0-6.6 5.4-12 12-12h32c6.6 0 12 5.4 12 12v141.7l66.8 48.6c5.4 3.9 6.5 11.4 2.6 16.8L334.6 349c-3.9 5.3-11.4 6.5-16.8 2.6z"
                fill={4289309613}
                fillOpacity={1}
                propList={
                  Array [
                    "fill",
                    "fillOpacity",
                  ]
                }
              />
            </RNSVGGroup>
          </RNSVGSvgView>
          <Text
            style={
              Array [
                Object {
                  "color": "#A9ABAD",
                  "display": "flex",
                  "fontFamily": "SFUIDisplay-Regular",
                  "fontSize": 14,
                  "lineHeight": 20,
                },
              ]
            }
          >
            25
             mins
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
